### PR TITLE
Pass environment variable to allow file transport

### LINKED ISSUE #5896: GitRepositoryTests fail with recent versions of git
### Description

Due to a [security issue](https://github.com/git/git/security/advisories/GHSA-3wp6-j8xr-qw85), Git no longer allows access to the `file` protocol by default.

Two of the test cases (`testSubmoduleRead` and `testSubmodules`) in `GitRepositoryTests` fail because the submodules are using a local file path:
```
a47s1v9v026q46t1400glq4/T/spm-tests-testSubmodules.wfMinO/bar-original submodule add /var/folders/lg/jyk3lrvd47s1v9v026q46t1400glq4/T/spm-tests-testSubmodules.wfMinO/baz-original baz output:
    Cloning into '/private/var/folders/lg/jyk3lrvd47s1v9v026q46t1400glq4/T/spm-tests-testSubmodules.wfMinO/bar-original/baz'...
    fatal: transport 'file' not allowed
    fatal: clone of '/var/folders/lg/jyk3lrvd47s1v9v026q46t1400glq4/T/spm-tests-testSubmodules.wfMinO/baz-original' into submodule path '/private/var/folders/lg/jyk3lrvd47s1v9v026q46t1400glq4/T/spm-tests-testSubmodules.wfMinO/bar-original/baz' failed
```

### Expected behavior

The tests should pass.

### Actual behavior

The tests failed with an error.

### Steps to reproduce

1. Install a version of git with the security fix: >= v2.30.6, v2.31.5, v2.32.4, v2.33.5, v2.34.5, v2.35.5, v2.36.3, v2.37.4, v2.38.1
2. Run the tests.

### Swift Package Manager version/commit hash

_No response_

### Swift & OS version (output of `swift --version && uname -a`)

swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.135.3 clang-1400.0.29.51)
Target: x86_64-apple-macosx12.0
Darwin ip-address.redacted.com 21.6.0 Darwin Kernel Version 21.6.0: Thu Sep 29 20:12:57 PDT 2022; root:xnu-8020.240.7~1/RELEASE_X86_64 x86_64

### PR DESCRIPTION
**Pull request by dtweston**
This change fixes #5896, by explicitly allowing the `file` transport for the tests that require it.

### Motivation:

Recent versions of git made a security fix to disallow the `file` transport by default. The `GitRepository` submodule test cases fail with these recent versions, because they use local submodules which rely on the `file` transport.

### Modifications:

This change makes it possible to pass in a set of environment variables when creating the `GitRepositoryProvider`, using the default `Git.environment` if not provided in the initializer. These environment variables are used for all subsequent git commands unless overridden for a particular `GitShellHelper.run(_:environment:outputRedirection:)` invocation (which is a capability that doesn't seem to be used currently).

The `GitRepositoryTests` submodule tests set the environment variable `GIT_ALLOW_PROTOCOL=file` to tell git to allow use of the `file` protocol only for these tests where it's required.

### Result:

The submodule tests will pass because they will be explicitly enabling the `file` transport.

### CODE REVIEW

**Comment by tomerd** [comment]
I wonder if the "better" fix here is to change the tests not to use the `file://` prefix. @neonichu @abertelrud wdyt?

**Comment by neonichu** [comment]
> I wonder if the "better" fix here is to change the tests not to use the file:// prefix. @neonichu @abertelrud wdyt?

Is there an alternative for a local Git repo? I'm not sure I fully understand what it means that the file transport is unsupported by default now.

**Comment by dtweston** [comment]
The tests don't explicitly use a file schema, it's just because they use a local path. 

The scenario presented in the [security bulletin](https://github.com/git/git/security/advisories/GHSA-3wp6-j8xr-qw85) is that an attacker could somehow get access to a sensitive file on a user's workstation by creating a malicious repo and getting them to clone it. 

You can still clone local repos manually from the command line, as that's determined to be a user action, but the implicit clones that happen as part of adding or updating a submodule is disallowed, unless overridden. It's possible to override it via ` git config --global`, but that should be left up to the end user. 
Adding the environment variable tells git, "it's ok, I trust the source of these local repos you're cloning on my behalf".

As for alternatives to the
`file` transport... you could set up something that hosts the submodule over `https` or `ssh`, but that sounds like way more trouble than it's worth.

**Comment by neonichu** [comment]
Thanks for the explanation, @dtweston. I think that sounds as if allowing the file transport for these tests is the right thing to do.

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
thanks for the explanation @dtweston 

> You can still clone local repos manually from the command line, as that's determined to be a user action, but the implicit clones that happen as part of adding or updating a submodule is disallowed, unless overridden.

how does this impact SwiftPM more broadly, or iow, how are the tests different from real-world use cases where SwiftPM needs to work with local repositories?

**Comment by neonichu** [comment]
Are there any real world use cases for using a local repository? I have only ever seen them be used for tests / experimentation.

**Comment by tomerd** [comment]
> Are there any real world use cases for using a local repository with SwiftPM? I have only ever seen them be used for tests / experimentation.

if that is the case, do we actually want to support them? ie if they are test only feature maybe we refactor the code so its not possible use outside tests?

**Comment by dtweston** [comment]
There might be real world use cases, but that isn't SwiftPM's problem. If a particular user has a workflow that requires local submodules, they'll need to configure their environment to support it.

The way the change is written, the extra environment variable is only passed by the tests that need it. The production code already had the ability to override environment variables for a particular git call. The only change is that the default environment variables can be set at a higher level (`GitRepositoryProvider`) so it gets used by default for any repositories created by it.

File: Tests/SourceControlTests/GitRepositoryTests.swift [hunk 1]
```diff
@@ -563,15 +567,18 @@ class GitRepositoryTests: XCTestCase {
...
             try fooWorkingRepo.checkout(tag: "1.0.0")
             XCTAssertNoSuchPath(fooWorkingPath.appending(component: "bar"))
 
             // Add submodule to foo and tag it as 1.0.1
             try foo.checkout(newBranch: "submodule")
-            try systemQuietly([Git.tool, "-C", fooPath.pathString, "submodule", "add", barPath.pathString, "bar"])
+
+            try Process.checkNonZeroExit(
+                args: Git.tool, "-C", fooPath.pathString, "submodule", "add", barPath.pathString, "bar",
+                environment: ["GIT_ALLOW_PROTOCOL": "file"])
```
Comment (tomerd) [inline Tests/SourceControlTests/GitRepositoryTests.swift:581]: use testEnvironment here?

File: Tests/SourceControlTests/GitRepositoryTests.swift [hunk 2]
```diff
@@ -589,7 +596,9 @@ class GitRepositoryTests: XCTestCase {
             // Add something to bar.
             try localFileSystem.writeFileContents(barPath.appending(component: "bar.txt"), bytes: "hello")
             // Add a submodule too to check for recursive submodules.
-            try systemQuietly([Git.tool, "-C", barPath.pathString, "submodule", "add", bazPath.pathString, "baz"])
+            try Process.checkNonZeroExit(
+                args: Git.tool, "-C", barPath.pathString, "submodule", "add", bazPath.pathString, "baz",
+                environment: ["GIT_ALLOW_PROTOCOL": "file"])
```
Comment (tomerd) [inline Tests/SourceControlTests/GitRepositoryTests.swift:601]: use testEnvironment here?

**Comment by tomerd** [comment]
> The only change is that the default environment variables can be set at a higher level (GitRepositoryProvider) so it gets used by default for any repositories created by it.

right, I am trying to see if there is amore minimal change we can make to address this

**Comment by dtweston** [comment]
Hmmmm, I did look at modifying `Git.environment` first, but it's a static var that's defined in `swift-tools-support-core`, so I wasn't sure what else might be affected by changes there, and I didn't know of a way for it to vary its behavior just for tests.

I'm happy to try something different if you have suggestions for the second part.

**Comment by tomerd** [comment]
here is a proposal:
https://github.com/apple/swift-tools-support-core/pull/371
https://github.com/apple/swift-package-manager/pull/5914

**Comment by tomerd** [comment]
closing in favor or https://github.com/apple/swift-package-manager/pull/5914

**Comment by tomerd** [comment]
thank you for helping with this @dtweston

**Pull request closed without merging**